### PR TITLE
Use masking for backward operation

### PR DESCRIPTION
**Pull request by jadore801120**
Hello, thanks for this awesome repo!
Follow the TODO comment in the code, I rewrite the backward part with masking.
Please take a look.
Thanks.
Yu-Hsiang

### CODE REVIEW

File: vq_vae/nearest_embed.py [hunk 1]
```diff
@@ -50,15 +50,15 @@ def backward(ctx, grad_output, argmin=None):
...
-            grad_output_reshaped = grad_output.permute(0, *ctx.dims[2:], 1).contiguous().view(ctx.batch_size * ctx.num_latents, ctx.emb_dim)
-            argmin_flat = ctx.argmin.view(-1)
-            # TODO: replace for loop, use masked operations
-            for i in range(ctx.num_emb):
-                if torch.sum(ctx.argmin == i):
-                    indices_chosen_atom_i = torch.nonzero(argmin_flat == i)
-                    gradients_of_latents_chosen_atom_i = grad_output_reshaped[indices_chosen_atom_i, :]
-                    grad_emb[:, i] = torch.mean(gradients_of_latents_chosen_atom_i, 0)
+            latent_indices = torch.arange(ctx.num_emb).type_as(ctx.argmin)
+            idx_choices = (ctx.argmin.view(-1, 1) == latent_indices.view(1, -1)).float()
```
Comment (nadavbh12) [inline vq_vae/nearest_embed.py]: `.type_as(grad_output.data)` is more appropriate than `float()`.

**Review by nadavbh12 — COMMENTED** [review]
Thanks! Just one small comment...

**Comment by jadore801120** [comment]
Hi, thanks for the review! I did not notice that before.
I have fixed it, please take a look.
